lib/minurl.browser.js

```diff
@@ -0,0 +1,83 @@
+/// <reference lib="dom" />
+
+// See: <https://github.com/nodejs/node/blob/fcf8ba4/lib/internal/url.js>
+
+/**
+ * @param {string|URL} path
+ */
+export function urlToPath(path) {
+  if (typeof path === 'string') {
+    path = new URL(path)
+  } else if (!isUrl(path)) {
+    /** @type {NodeJS.ErrnoException} */
+    const error = new TypeError(
+      'The "path" argument must be of type string or an instance of URL. Received `' +
+        path +
+        '`'
+    )
+    error.code = 'ERR_INVALID_ARG_TYPE'
+    throw error
+  }
+
+  if (path.protocol !== 'file:') {
+    /** @type {NodeJS.ErrnoException} */
+    const error = new TypeError('The URL must be of scheme file')
+    error.code = 'ERR_INVALID_URL_SCHEME'
+    throw error
+  }
+
+  return getPathFromURLPosix(path)
+}
+
+/**
+ * @param {URL} url
+ */
+function getPathFromURLPosix(url) {
+  if (url.hostname !== '') {
+    /** @type {NodeJS.ErrnoException} */
+    const error = new TypeError(
+      'File URL host must be "localhost" or empty on darwin'
+    )
+    error.code = 'ERR_INVALID_FILE_URL_HOST'
+    throw error
+  }
+
+  const pathname = url.pathname
+  let index = -1
+
+  while (++index < pathname.length) {
+    if (
+      pathname.charCodeAt(index) === 37 /* `%` */ &&
+      pathname.charCodeAt(index + 1) === 50 /* `2` */
+    ) {
+      const third = pathname.charCodeAt(index + 2)
+      if (third === 70 /* `F` */ || third === 102 /* `f` */) {
+        /** @type {NodeJS.ErrnoException} */
+        const error = new TypeError(
+          'File URL path must not include encoded / characters'
+        )
+        error.code = 'ERR_INVALID_FILE_URL_PATH'
+        throw error
+      }
+    }
+  }
+
+  return decodeURIComponent(pathname)
+}
+
+// Note: this is copy/pasted from `minurl.js` because otherwise DOM/Node types
+// mix up.
+/**
+ * @param {unknown} fileURLOrPath
+ * @returns {fileURLOrPath is URL}
+ */
+export function isUrl(fileURLOrPath) {
+  return (
+    fileURLOrPath !== null &&
+    typeof fileURLOrPath === 'object' &&
+    // @ts-expect-error: indexable.
+    fileURLOrPath.href &&
+    // @ts-expect-error: indexable.
+    fileURLOrPath.origin
+  )
+}
```

Comment:
if it is the same in both, would it make sense to extract to another file shared between the two?

Reply:
I tried but it results in type errors because the one `URL` is not assignable to the other `URL` 🤷‍♂️

Reply:
Previous approach did leak node types. I’ve now “manually typed” URL, which prevents that from happening.